File: sdk/python/examples/submit_pr.py
#!/usr/bin/env python3
"""
Example: Submit a Pull Request

This example shows how to submit a PR and send a review request message.
"""

from eket_sdk import EketClient, AgentType, AgentRole, MessageType, TestStatus

# Initialize and register
client = EketClient(server_url="http://localhost:8080")
agent = client.register_agent(
    agent_type=AgentType.CUSTOM,
    role=AgentRole.SLAVER,
)

print(f"✅ Registered as {agent.instance_id}")

# Assume we've completed a task
TASK_ID = "FEAT-001"  # Replace with actual task ID
BRANCH = f"feature/{TASK_ID}-user-authentication"

print(f"\n📤 Submitting PR for task {TASK_ID}...")

# Submit PR
pr_id = client.submit_pr(
    instance_id=agent.instance_id,
    task_id=TASK_ID,
    branch=BRANCH,
    description="""
## Summary
Implemented user authentication with JWT tokens.

## Changes
- Added login/logout endpoints
- Implemented JWT token generation and validation
- Added password hashing with bcrypt
- Created user session management

## Test Results
- Unit tests: 42/42 passed
- Integration tests: 15/15 passed
- Coverage: 87%

## Checklist
- [x] Code follows style guidelines
- [x] Tests added and passing
- [x] Documentation updated
- [x] No breaking changes
    """,
    test_status=TestStatus.PASSED,
)

print(f"   ✅ PR submitted: {pr_id}")

# Send review request message to master
print(f"\n📨 Sending review request to master...")

# Find master agent
masters = client.list_agents(role=AgentRole.MASTER)
if not masters:
    print("   ⚠️  No master agents available")
else:
    master_id = masters[0].instance_id
    print(f"   Sending to: {master_id}")

    message_id = client.send_message(
        from_id=agent.instance_id,
        to_id=master_id,
        msg_type=MessageType.PR_REVIEW_REQUEST,
        payload={
            "task_id": TASK_ID,
            "branch": BRANCH,
            "pr_id": pr_id,
            "description": "User authentication feature ready for review",
            "test_status": "passed",
            "test_coverage": 0.87,
            "files_changed": 12,
            "insertions": 450,
            "deletions": 23,
        },
    )

    print(f"   ✅ Message sent: {message_id}")

print(f"\n✅ PR submission complete!")
print(f"   Task: {TASK_ID}")
print(f"   Branch: {BRANCH}")
print(f"   Status: Awaiting review")

# Clean up
client.close()

File: sdk/javascript/examples/submit-pr.ts
/**
 * Example: PR Review and Merge (Master Role)
 *
 * This example demonstrates the master workflow:
 * 1. Register as master
 * 2. Connect to WebSocket for real-time notifications
 * 3. List agents and tasks
 * 4. Review a pull request
 * 5. Merge approved PR
 */

import { EketClient, Message } from '../src/index.js';

async function main() {
  const client = new EketClient({
    serverUrl: 'http://localhost:8080',
    enableWebSocket: true,
  });

  try {
    // Step 1: Register as master
    console.log('Step 1: Registering as master...');
    const { instance_id } = await client.registerAgent({
      agent_type: 'claude_code',
      role: 'master',
      capabilities: ['code_review', 'architecture', 'coordination'],
    });
    console.log('✓ Registered as master:', instance_id);

    // Step 2: Connect WebSocket for real-time notifications
    console.log('\nStep 2: Connecting to WebSocket...');
    await client.connectWebSocket(instance_id);
    console.log('✓ WebSocket connected');

    // Set up message handler
    client.onMessage((message: Message) => {
      console.log(`\n📧 Received message: ${message.type} from ${message.from}`);
      if (message.type === 'pr_review_request') {
        const { task_id, description } = message.payload as any;
        console.log(`  PR review requested for task: ${task_id}`);
        console.log(`  Description: ${description}`);
      }
    });

    // Step 3: List agents and tasks
    console.log('\nStep 3: Checking system status...');
    const agents = await client.listAgents({ status: 'active' });
    console.log(`✓ Active agents: ${agents.length}`);
    agents.forEach((agent) => {
      console.log(`  - ${agent.instance_id} (${agent.role}/${agent.specialty})`);
    });

    const reviewTasks = await client.listTasks({ status: 'review' });
    console.log(`\n✓ Tasks in review: ${reviewTasks.length}`);

    if (reviewTasks.length === 0) {
      console.log('ℹ No tasks to review. Exiting...');
      await client.shutdown();
      return;
    }

    // Step 4: Review a PR
    const taskToReview = reviewTasks[0];
    console.log(`\nStep 4: Reviewing PR for task: ${taskToReview.id}`);
    console.log(`  Title: ${taskToReview.title}`);
    console.log(`  Assigned to: ${taskToReview.assigned_to}`);

    // Fetch task details
    const task = await client.getTask(taskToReview.id);
    console.log('  Description:', task.description);

    // Submit review
    console.log('\n  Submitting review...');
    await client.reviewPR(task.id, {
      reviewer: instance_id,
      status: 'approved',
      comments: [
        {
          file: 'src/main.ts',
          line: 42,
          comment: 'Great implementation!',
        },
      ],
      summary: 'Code looks good. Approving for merge.',
    });
    console.log('  ✓ Review submitted: APPROVED');

    // Step 5: Merge PR
    console.log('\nStep 5: Merging PR...');
    const mergeResult = await client.mergePR(task.id, {
      merger: instance_id,
      target_branch: 'main',
      squash: false,
    });
    console.log('✓ PR merged successfully');
    console.log('  Merge commit:', mergeResult.merge_commit);
    console.log('  Merged at:', mergeResult.merged_at);

    // Update task to done
    console.log('\nStep 6: Marking task as done...');
    await client.updateTask(task.id, {
      status: 'done',
    });
    console.log('✓ Task marked as done');

    // Send notification to assignee
    if (task.assigned_to) {
      console.log('\nStep 7: Notifying assignee...');
      await client.sendMessage({
        from: instance_id,
        to: task.assigned_to,
        type: 'pr_approved',
        priority: 'normal',
        payload: {
          task_id: task.id,
          message: `Your PR for ${task.title} has been approved and merged!`,
          merge_commit: mergeResult.merge_commit,
        },
      });
      console.log('✓ Assignee notified');
    }

    // Keep connection alive for a bit to receive any messages
    console.log('\nListening for messages (5 seconds)...');
    await new Promise((resolve) => setTimeout(resolve, 5000));

    // Cleanup
    console.log('\nCleaning up...');
    await client.shutdown();
    console.log('✓ Shutdown complete');

    console.log('\n✅ PR review and merge workflow completed!');
  } catch (error) {
    console.error('❌ Error:', error);
    process.exit(1);
  }
}

main();
